examples/shopping_ads/add_performance_max_product_listing_group_tree.rb

```diff
@@ -0,0 +1,481 @@
+    @next_id = @root_listing_group_id - 1
+  end
+
+  # Returns a new temporary ID to be used for a resource name in a
+  # MutateOperation. See
+  # https://developers.google.com/google-ads/api/docs/mutating/best-practices#temporary_resource_names
+  # for details about temporary IDs.
+  def next_id
+    @next_id -= 1
+  end
+
+  # [START add_performance_max_product_listing_group_tree_4]
+  # Creates a MutateOperation that creates a root AssetGroupListingGroupFilter
+  # for the factory's AssetGroup.
+  #
+  # The root node or partition is the default, which is displayed as "All
+  # Products".
+  def create_root(client)
+    operation = client.operation.create_resource.asset_group_listing_group_filter do |lgf|
+      lgf.resource_name = client.path.asset_group_listing_group_filter(
+        @customer_id,
+        @asset_group_id,
+        @root_listing_group_id,
+      )
+      lgf.asset_group = client.path.asset_group(
+        @customer_id,
+        @asset_group_id,
+      )
+
+      # Since this is the root node, do not set the ParentListingGroupFilter.
+      # For all other nodes, this would refer to the parent listing group
+      # filter resource name.
+      # lgf.parent_listing_group_filter = "<PARENT FILTER NAME>"
+
+      # Unlike AddPerformanceMaxRetailCampaign, the type for the root node here
+      # must be Subdivision because we add child partitions under it.
+      lgf.type = :SUBDIVISION
+
+      # Because this is a Performance Max campaign for retail, we need to
+      # specify that this is in the shopping vertical.
+      lgf.vertical = :SHOPPING
+    end
+
+    client.operation.mutate do |m|
+      m.asset_group_listing_group_filter_operation = operation
+    end
+  end
+  # [END add_performance_max_product_listing_group_tree_4]
+
+  # [START add_performance_max_product_listing_group_tree_5]
+  # Creates a MutateOperation that creates a intermediate
+  # AssetGroupListingGroupFilter for the factory's AssetGroup.
+  #
+  # Use this method if the filter will have child filters. Otherwise, use the
+  # create_unit method.
+  def create_subdivision(client, parent, id, dimension)
+    operation = client.operation.create_resource.asset_group_listing_group_filter do |lgf|
+      lgf.resource_name = client.path.asset_group_listing_group_filter(
+        @customer_id,
+        @asset_group_id,
+        id,
+      )
+      lgf.asset_group = client.path.asset_group(
+        @customer_id,
+        @asset_group_id,
+      )
+      lgf.parent_listing_group_filter = client.path.asset_group_listing_group_filter(
+        @customer_id,
+        @asset_group_id,
+        parent,
+      )
+
+      # We must use the Subdivision type to indicate that the
+      # AssetGroupListingGroupFilter will have children.
+      lgf.type = :SUBDIVISION
+
+      # Because this is a Performance Max campaign for retail, we need to
+      # specify that this is in the shopping vertical.
+      lgf.vertical = :SHOPPING
```

Comment:
nit: In Java, we change Subdivision and shopping in the comments to all caps so they match the enum naming convention for the language.

Reply:
Done